core/src/main/java/org/apache/accumulo/core/iterators/user/SeekingFilter.java

```diff
@@ -0,0 +1,221 @@
+
+    public FilterResult(boolean accept, AdvanceResult advance) {
+      this.accept = accept;
+      this.advance = advance;
+    }
+
+    public static FilterResult of(boolean accept, AdvanceResult advance) {
+      return accept ? PASSES.get(advance) : FAILS.get(advance);
+    }
+
+    public String toString() {
+      return "Acc: " + accept + " Adv: " + advance;
+    }
+  }
+
+  /**
+   * Subclasses must provide an implementation which examines the given key and value and determines (1) whether to accept the KV pair and (2) how far to
+   * advance the source iterator past the key.
+   *
+   * @param k
+   *          a key
+   * @param v
+   *          a value
+   * @return indicating whether to pass or block the key, and how far the source iterator should be advanced.
+   */
+  public abstract FilterResult filter(Key k, Value v);
+
+  /**
+   * Whenever the subclass returns AdvanceResult.USE_HINT from its filter predicate, this method will be called to see how far to advance the source iterator.
+   * The return value must be a key which is greater than (sorts after) the input key. If the subclass never returns USE_HINT, this method will never be called
+   * and may safely return null.
+   *
+   * @param k
+   *          a key
+   * @param v
+   *          a value
+   * @return as above
+   */
+  public abstract Key getNextKeyHint(Key k, Value v);
+
+  private Collection<ByteSequence> columnFamilies;
+  private boolean inclusive;
+  private Range seekRange;
+  private boolean negate;
+
+  private AdvanceResult advance;
+
+  private boolean advancedPastSeek = false;
+
+  @Override
+  public void next() throws IOException {
+    findTop();
+  }
+
+  @Override
+  public boolean hasTop() {
+    return !advancedPastSeek && super.hasTop();
+  }
+
+  @Override
+  public void seek(Range range, Collection<ByteSequence> columnFamilies, boolean inclusive) throws IOException {
+    super.seek(range, columnFamilies, inclusive);
+    advance = null;
+    this.columnFamilies = columnFamilies;
+    this.inclusive = inclusive;
+    seekRange = range;
+    advancedPastSeek = false;
+    findTop();
+  }
+
+  @Override
+  public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+    super.init(source, options, env);
+    negate = Boolean.parseBoolean(options.get(NEGATE));
+  }
+
+  @Override
+  public SortedKeyValueIterator<Key,Value> deepCopy(IteratorEnvironment env) {
+    SeekingFilter newInstance;
+    try {
+      newInstance = this.getClass().newInstance();
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+    newInstance.setSource(getSource().deepCopy(env));
+    newInstance.negate = negate;
+    return newInstance;
+  }
+
+  protected void findTop() throws IOException {
+    SortedKeyValueIterator<Key,Value> src = getSource();
+    // advance could be null if we've just been seeked
+    if (src.hasTop() && advance != null) {
```

Comment:
Good point. Fixed in 71b1b20. In `next()`, `advance` should never be null and no need to check `src.hasTop()` because `hasTop` already checks that, and calling `next` when `hasTop` is false is a violation of the iterator contract.